src/zcash/circuit/gadget.tcc

```
@@ -0,0 +1,349 @@
+        // the intended root in the event that
+        // both inputs are zero-valued.
+        zk_merkle_root->bits.fill_with_bits(
+            this->pb,
+            uint256_to_bool_vector(rt)
+        );
+
+        // Witness public balance values
+        zk_vpub_old.fill_with_bits(
+            this->pb,
+            uint64_to_bool_vector(vpub_old)
+        );
+        zk_vpub_new.fill_with_bits(
+            this->pb,
+            uint64_to_bool_vector(vpub_new)
+        );
+
+        {
+            // Witness total_uint64 bits
+            uint64_t left_side_acc = vpub_old;
+            for (size_t i = 0; i < NumInputs; i++) {
+                left_side_acc += inputs[i].note.value;
+            }
+
+            zk_total_uint64.fill_with_bits(
+                this->pb,
+                uint64_to_bool_vector(left_side_acc)
+            );
+        }
+
+        // Witness phi
+        zk_phi->bits.fill_with_bits(
+            this->pb,
+            trailing252(uint256_to_bool_vector(phi))
+        );
+
+        // Witness h_sig
+        zk_h_sig->bits.fill_with_bits(
+            this->pb,
+            uint256_to_bool_vector(h_sig)
+        );
+
+        for (size_t i = 0; i < NumInputs; i++) {
+            // Witness the input information.
+            auto merkle_path = inputs[i].witness.path();
+            zk_input_notes[i]->generate_r1cs_witness(
+                merkle_path,
+                inputs[i].key,
+                inputs[i].note
+            );
+
+            // Witness macs
+            zk_mac_authentication[i]->generate_r1cs_witness();
+        }
+
+        for (size_t i = 0; i < NumOutputs; i++) {
+            // Witness the output information.
+            zk_output_notes[i]->generate_r1cs_witness(outputs[i]);
+        }
+
+        // [SANITY CHECK] Ensure that the intended root
+        // was witnessed by the inputs, even if the read
+        // gadget overwrote it. This allows the prover to
```

I'm mildly disturbed that anything can get overwritten in an API that should have been pure-functional, and I still don't quite understand this. I guess that making the prover fail early is not essential to security, though, so I'm fine with this for now.